src/source-control/GitSourceControl.ts

```diff
@@ -38,49 +53,396 @@ export class GitSourceControl implements SourceControl {
+          parentCommits: [targetCommitSttack.hash],
+          childCommits: [...childrenOfTargetCommit],
+        };
+        draftCommitHashMap.set(newCommitOid.tostrS(), newCommitToInsert);
+        // Update ChildCommitHash for targetCommitSttack to newCommit
+        targetCommitSttack.childCommits = [newCommitOid.tostrS()];
+        // Update ParentCommitHash for all of the children and replace targetCommitSttack's hash with newCommit's hash
+        childrenOfTargetCommit.forEach((child) => {
+          const childCommit = draftCommitHashMap.get(child)!;
+          const parentsOfChild = childCommit.parentCommits.map((parent) => {
+            if (parent === targetCommitSttack.hash) {
+              return newCommitOid.tostrS();
+            }
+            return parent;
+          });
+          childCommit.parentCommits = parentsOfChild;
+          draftCommitHashMap.set(child, childCommit);
+        });
+      });
+    }
+    // Remove temp branch
+    nodegit.Branch.delete(await repo.getBranch(tempBranchName));
+    await this.loadIfChangesPresent();
   }
 
   async pushCommit(commit: Commit): Promise<void> {
-    // TODO: Implement
+    try {
+      const {
+        userPublicKeyPath,
+        userPrivateKeyPath,
+        userPassphrase,
+      } = JSON.parse(
+        fs.readFileSync(`${this.repoPath}/sttack.config.json`, "utf-8"),
+      );
+      const repo = await nodegit.Repository.open(this.repoPath);
+      // SEE: Could be any of the long lived branches?
+      const remote = await repo.getRemote("origin");
+      const callback = {
+        credentials: function (_url: string, userName: string) {
+          //console.log("Getting credentials");
+          //   // FIXME: Possible infinite loop when using sshKeyFromAgent
+          //   // See: https://github.com/nodegit/nodegit/issues/1133
+          //   //return nodegit.Cred.sshKeyFromAgent(userName);
+          return nodegit.Cred.sshKeyNew(
+            userName,
+            userPublicKeyPath, //"/Users/phuonganh/.ssh/id_rsa.pub",
+            userPrivateKeyPath, //"/Users/phuonganh/.ssh/id_rsa",
+            userPassphrase, //"hello",
+          );
+        },
+      };
+      const connection = remote.connect(nodegit.Enums.DIRECTION.PUSH, callback);
+      // SEE: Can only push if branch exists on remote? How to find branch name here?
+      await Promise.all(
+        commit.refNames.map(
+          async (ref: string): Promise<number> => {
+            return await remote.push([`${ref}:${ref}`], {
+              callbacks: {
+                credentials: function (_: string, userName: string) {
+                  return nodegit.Cred.sshKeyNew(
+                    userName,
+                    userPublicKeyPath, //"/Users/phuonganh/.ssh/id_rsa.pub",
+                    userPrivateKeyPath, //"/Users/phuonganh/.ssh/id_rsa",
+                    userPassphrase, //"hello",
+                  );
+                },
+              },
```

Comment:
Could we reuse `callback` from above?

```suggestion
              callbacks: callback,
```

Reply:
Sure! Will do this